app/cluster/dynamic.go

```diff
@@ -0,0 +1,104 @@
+package cluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rudderlabs/rudder-server/utils/types/servermode"
+)
+
+type modeProvider interface {
+	ServerMode() <-chan servermode.ModeAck
+}
+
+type Lifecycle interface {
+	Start()
+	Stop()
+}
+
+type Dynamic struct {
+	Provider modeProvider
+
+	GatewayDB     Lifecycle
+	RouterDB      Lifecycle
+	BatchRouterDB Lifecycle
+	ErrorDB       Lifecycle
+
+	Processor Lifecycle
+	Router    Lifecycle
+
+	currentMode servermode.Mode
+}
+
+func (d *Dynamic) Run(ctx context.Context) error {
```

Comment:
Start with degraded mode instead of undefined.

Reply:
done